festim/hydrogen_transport_problem.py

```diff
@@ -502,18 +527,18 @@ def create_formulation(self):
                 )
                 self.formulation += ((u - u_n) / self.dt) * v * self.dx(vol.id)
 
-                # add sources
-                # TODO implement this
-                # for source in self.sources:
-                #     # f = Constant(my_mesh.mesh, (PETSc.ScalarType(0)))
-                #     if source.species == spe:
-                #         formulation += source * v * self.dx
-                # add fluxes
-                # TODO implement this
-                # for bc in self.boundary_conditions:
-                #     pass
-                #     if bc.species == spe and bc.type != "dirichlet":
-                #         formulation += bc * v * self.ds
+        # add sources
+        for source in self.sources:
+            self.formulation -= (
+                source.value_fenics * source.species.test_function * self.dx(vol.id)
```

Comment:
should it not be `self.dx(source.volume.id)` instead?

Reply:
indeed